### DISCUSSION TITLE
Question about graph sage and role of message_and_aggregate function

### DISCUSSION THREAD

**Discussion by gebrahimi91** [pyg-team/pytorch_geometric#2410] [Q&A]
For GraphSage, there is this part in the source code:

     def message_and_aggregate(self, adj_t: SparseTensor,
                              x: OptPairTensor) -> Tensor:

        adj_t = adj_t.set_value(None, layout=None)
        return matmul(adj_t, x[0], reduce=self.aggr)

Can i ask what is adj_t here? and why x is OptPairTensor? I also do not know when and how this function is executed. I tried to print adj_t, but it does not do anything for me. Even printing a random sentence does not happen. May I know how the propagate function executes the message and message_and_aggregate function? Thanks

**Comment by rusty1s** [comment]
Should be hopefully already answered by https://github.com/rusty1s/pytorch_geometric/discussions/2412 :)
